source/dofs/dof_handler_policy.cc

```diff
@@ -845,14 +845,108 @@ namespace internal
         template <int spacedim>
         static
         void
-        renumber_mg_dofs (const std::vector<dealii::types::global_dof_index> &,
-                          const IndexSet &,
-                          DoFHandler<3,spacedim> &,
-                          const unsigned int               ,
-                          const bool                       )
+        renumber_mg_dofs (const std::vector<dealii::types::global_dof_index> &new_numbers,
+                          const IndexSet &indices,
+                          DoFHandler<3,spacedim> &dof_handler,
+                          const unsigned int level,
+                          const bool check_validity)
         {
-          // TODO
-          AssertThrow(false, ExcNotImplemented());
+          if (level>=dof_handler.get_tria().n_levels())
+            return;
+          for (typename std::vector<typename DoFHandler<3,spacedim>::MGVertexDoFs>::iterator i=dof_handler.mg_vertex_dofs.begin();
+               i!=dof_handler.mg_vertex_dofs.end(); ++i)
+            // if the present vertex lives on
+            // the present level
+            if ((i->get_coarsest_level() <= level) &&
+                (i->get_finest_level() >= level))
+              for (unsigned int d=0; d<dof_handler.get_fe().dofs_per_vertex; ++d)
+                {
+                  dealii::types::global_dof_index idx =i->get_index (level, d);
+                  if (idx != DoFHandler<3>::invalid_dof_index)
+                    i->set_index (level, d,
```

Comment:
You are here first enumerating all of the vertex dofs on each level, then lines, etc. This means that the dof indices on every cell are not close to contiguous. Is this what we already do for the non-parallel case?